### ISSUE TITLE
Support `notion uninstall` for node and yarn

### ISSUE THREAD

**Issue by mikrostew** [notion-cli/notion#327]
Similar to #319 - since we have `notion install` for Node and Yarn, we should also support `notion uninstall`.

**Comment by chriskrycho** [comment]
Per conversation in Discord today (see esp. [1], [2], [3], [4]), this is probably worth designing and implementing alongside the logic for `volta install npm`. I've marked it for now as [<b>needs RFC</b>][label] accordingly.

[1]: https://discordapp.com/channels/413753499835301908/413753499835301917/576061595814461450
[2]: https://discordapp.com/channels/413753499835301908/413753499835301917/576062745779109908
[3]: https://discordapp.com/channels/413753499835301908/413753499835301917/576086087072612363
[4]: https://discordapp.com/channels/413753499835301908/413753499835301917/576088824132927498
[label]: https://github.com/notion-cli/notion/issues?q=is%3Aissue+is%3Aopen+label%3A%22needs+RFC%22

**Comment by BProg** [comment]
In a little discord [discussion](https://discordapp.com/channels/413753499835301908/413753499835301917/751459790453997570) around the possibility to _uninstall_ yarn and node, a little idea came up: 
> I was thinking that maybe they will get deleted automatically after some time of not using them xD

maybe worth trying 🤷

**Comment by dwiyatci** [comment]
Hi all 👋

is there an alternative way to do this manually while waiting for the RFC to be done? For example, if I'm not using `yarn@1.22.10` anymore and want to "uninstall" it, would it be enough to `rm -rf` the followings?

- `~/.volta/tools/image/yarn/1.22.10`
- `~/.volta/tools/inventory/yarn/yarn-v1.22.10.tar.gz`

Thank you.

**Comment by charlespierce** [comment]
@dwiyatci Yep, that's exactly what you'd need to do to reclaim the space: The first directory is the unpacked install, while the 2nd is the cached tarball.

Possibly important caveat: After deleting, if you come across that version again in a project, it will be automatically re-downloaded (in the same way that they're automatically downloaded the first time they are seen).

**Comment by migs35323** [comment]
Recently node upped from 18.0.0 to 18.1.0 and i just "updated" it so i have no great reason to keep the  8.flat version, so i was looking to unninstall it just to don't have a massive list of leftovers, maybe based on "volta uninstall" we can also get a "volta prune"? so it deletes the minor releases?
should work with the tools and package managers as well
examples:
installed list:
- node 18.1.0
- node 18.0
- node 16.15.1
- node 16.14.2
- node 16.14.1
- node 16.12.0
- node 12.8
- node 12.5
$ volta prune node@16 -y
- node 18.1.0
- node 18.0
- node 16.15.1
- node 12.8
- node 12.5
$ volta prune node -y
- node 18.1.0
- node 16.15.1
- node 12.8

**Comment by hxii** [comment]
I've made a small utility to remove unused Node versions from Volta:
https://pypi.org/project/vnu/

**Comment by dwiyatci** [comment]
> I've made a small utility to remove unused Node versions from Volta: https://pypi.org/project/vnu/

@hxii I love the "Disclaimer" section 😃

**Comment by TheRealPSV** [comment]
For any windows users who end up here, the equivalent folder to `~/.volta/tools` in Windows is `%localappdata%\Volta\tools`

You can delete the mentioned folders in the same way.

**Comment by MrChocolatine** [comment]
Friendly bump, after 2+ years.

**Comment by NayamAmarshe** [comment]
No solution yet? Unfortunately, I will have to move away from Volta because of this. I do not want yarn v4 installed by default and I do not want to pin files in someone else's project either.

**Comment by charlespierce** [comment]
Hi @NayamAmarshe, can you clarify the trouble you’re running into? Volta doesn’t install anything by default: It runs what it determines is the appropriate version of a tool. It figures that out by looking in 2 places: Pinned in the current project or explicitly selected as the global default by you.

Volta will only use Yarn 4 if you choose that version (either as your default or if a project chooses it). Currently, there _is_ an issue where Yarn 4 is published as the `latest` version in the npm registry—see #1443—however you can always choose an explicit version when selecting your global default. For example, if you want to stick to using Yarn 1, you could run `volta install yarn@1`.

Lastly, the nature of Volta’s “fetch on-demand” model means that “uninstall” isn’t really the proper analogy for this command. It’s really closer to “cache clean” since the appropriate version will be redownloaded as necessary.

**Comment by NayamAmarshe** [comment]
> For example, if you want to stick to using Yarn 1, you could run `volta install yarn@1`.

Thank you, it's actually my fault that I forgot I could do `volta install yarn@1`, yarn v4 created a lot of problems for me. I'd still however like to request support with the uninstall command because it's just one of those things that improve the UX for the user.

```
$ volta uninstall yarn
error: Uninstalling yarn is not supported yet.
```

I think behind the scenes you could do anything but showing "Uninstalling yarn is not supported yet" makes it seem like Volta has missing features and is incomplete. So, that's the only reason I'd like to request this.

Thank you

**Comment by algoflows** [comment]
So my yarn is stuck at 3.x now and I can't uninstall it and installing yarn1 doesn't downgrade it to yarn 1.... volta what a mess.

**Comment by Konglomneshued** [comment]
It's unacceptable that there is no way to uninstall, especially after three freaking years.

**Comment by MrChocolatine** [comment]
4 years+ now, this issue was created on [27 March 2019](https://github.com/volta-cli/volta/issues/327#issue-426210008).

**Comment by ericp3reira** [comment]
Hey people!

I was trying to remove `yarn` and still not able to do it through Volta, so I removed the folders as mentioned in https://github.com/volta-cli/volta/issues/327#issuecomment-920336408.

Since Volta was still referencing `yarn` from somewhere, on top of what @dwiyatci mentioned, I also:
* removed the yarn version from `~/.volta/tools/user/platform.json`
* deleted the symbolic link from `~/.volta/bin` (mine was `yarn -> ~/.volta/bin/volta-shim`)

**Comment by ericp3reira** [comment]
> So my yarn is stuck at 3.x now and I can't uninstall it and installing yarn1 doesn't downgrade it to yarn 1.... volta what a mess.

@algoflows I was able to go back to v1 by removing all versions except the v1.22 (https://github.com/volta-cli/volta/issues/327#issuecomment-920336408) and `volta install yarn@1` later

**Comment by 1111mp** [comment]
Not being able to uninstall Node is really frustrating...but there is no doubt that Volta is still a very great project.

I wrote a `Node` version management tool myself: ([nvm-desktop](https://github.com/1111mp/nvm-desktop)), but it is a desktop application with a visual interface. Similar to `Volta`, but it only provides a shim for the `Node` engine and does not affect the behavior of the Node engine(npm,yarn,pnpm...) itself at all.

The download, installation and management of Node are performed through the client interface, and command line tools are also provided to manage different versions.

![image](https://github.com/volta-cli/volta/assets/31227919/db6ca864-83b2-4164-85fc-354f9a17ebbe)
![image](https://github.com/volta-cli/volta/assets/31227919/5c7af589-240e-41c1-9126-fdb4a04276be)

Hope it helps you.

**Comment by TLovers** [comment]
It's almost 2024

**Comment by chriskrycho** [comment]
It's true! And maybe in 2024 someone will fund the work to get this both designed and implemented; that would be awesome! 😅

**Comment by morishin** [comment]
Happy new year 2024! 🤚 

I couldn't understand why this has not been implemented for so many years, so I created a simple implementation #1658 for discussion.
I would like to know whether this approach is acceptable or if there are any concerns.

**Comment by NayamAmarshe** [comment]
> Happy new year 2024! 🤚 
> 
> 
> 
> I couldn't understand why this has not been implemented for so many years, so I created a simple implementation for discussion.
> 
> I would like to know whether this approach is acceptable or if there are any concerns.
> 
> 
> 
> - #1658

Yay! Finally! I hope they merge this! It's been 5 years!

**Comment by chriskrycho** [comment]
Thank you! At a basic level, the only reason it has not been implemented is that we’re all super busy and none of our employers has had the need for it so it has not gotten our time and attention (as you can see from looking at our activity level, it is “squeeze in maintenance in spare time because we care” mode… like me doing a bunch of triage on a Saturday morning!). I will look at that POC either later today or sometime early next week, but I am ***EXCITED!!!*** about the prospect of it landing.

**Comment by trry-hub** [comment]
<img width="383" alt="image" src="https://github.com/volta-cli/volta/assets/67309192/48d7eded-d1e6-4576-b925-067680ac0e68">

Do you have a solution now？

**Comment by MrChocolatine** [comment]
> Do you have a solution now？

@trry-hub 

https://github.com/volta-cli/volta/issues/327#issuecomment-920336408

**Comment by trry-hub** [comment]
> > Do you have a solution now？现在有解决办法了吗？
> 
> @trry-hub
> 
> [#327 (comment)](https://github.com/volta-cli/volta/issues/327#issuecomment-920336408)

This solution is too Muggle, I have already used NVM

**Comment by MrChocolatine** [comment]
@trry-hub

> This solution is too Muggle, I have already used NVM

- Too Muggle?
- You asked "Do you have a solution now?", I gave you a link to a solution to delete a version of Node.js.

PS: I did not write in Chinese in my original message.

**Comment by trry-hub** [comment]
> @trry-hub
> 
> > This solution is too Muggle, I have already used NVM
> 
> * Too Muggle?
> * You asked "Do you have a solution now?", I gave you a link to a solution to delete a version of Node.js. (since the CLI still does not include such feature, unfortunately)
> 
> PS: I did not write in Chinese in my original message.

I just don't think it's elegant enough. Thank you for your reply

**Comment by TheJaredWilcurt** [comment]
* `volta pin node@22.1.0` works
* `volta pin node@22.2.0` works
* `volta uninstall node@22.1.0` should work

hope it's added soon.

**Comment by mulder999** [comment]
I initially thought there was an issue with my Volta installation when I came across this thread. It seems that the `volta uninstall` command either should be removed if it's not implemented, or it should display a clearer message indicating that the command is not yet functional. The current error message, "No package node@22.3.0 found to uninstall," is completly misleading and should be revised.

**Comment by chriskrycho** [comment]
Yep, agreed. That message is fixed in the upcoming release! (When? Soon, hopefully!)

**Comment by ZinkWu** [comment]
Version 2.0.1 still does not support it.

**Comment by TheKevinWang** [comment]
been over 5 years and still nothing. pretty dumb to think of a install with no uninstall in the first place if you ask me.

**Comment by CreeTar** [comment]
we advise to not use volta anymore, cluttering your system without the ability to uninstall is just not common sense. please close this issue.

**Comment by yxw007** [comment]
Alternative solution: Clean up manually directly
- macOS/Linux: ~/.volta/tools/image/
- Windows: %LOCALAPPDATA%\Volta\tools\image\ (C:\Users\[UserName]\AppData\Local\Volta\tools\image\)
